clients/src/test/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManagerTest.java

```diff
@@ -455,6 +468,192 @@ public void testRequestFails_AuthenticationException() {
+
+        // No api version info initially available
+        when(apiVersions.get(LEADER_1.idString())).thenReturn(null);
+        requestManager.validatePositionsIfNeeded();
+        assertEquals(0, requestManager.requestsToSend(), "Invalid request count");
+        verify(subscriptionState, never()).completeValidation(TEST_PARTITION_1);
+        verify(subscriptionState, never()).setNextAllowedRetry(any(), anyLong());
+
+        // Api version updated, next validate positions should successfully build the request
+        when(apiVersions.get(LEADER_1.idString())).thenReturn(NodeApiVersions.create());
+        when(subscriptionState.partitionsNeedingValidation(time.milliseconds())).thenReturn(Collections.singleton(TEST_PARTITION_1));
+        when(subscriptionState.position(any())).thenReturn(position, position);
+        requestManager.validatePositionsIfNeeded();
+        assertEquals(1, requestManager.requestsToSend(), "Invalid request count");
+    }
+
+    private void expectSuccessfulBuildRequestForValidatingPositions(SubscriptionState.FetchPosition position, Node leader) {
+        when(subscriptionState.partitionsNeedingValidation(time.milliseconds())).thenReturn(Collections.singleton(TEST_PARTITION_1));
+        when(subscriptionState.position(any())).thenReturn(position, position);
+        NodeApiVersions nodeApiVersions = NodeApiVersions.create();
+        when(apiVersions.get(leader.idString())).thenReturn(nodeApiVersions);
+    }
+
+    private void testResetPositionsSuccessWithLeaderEpoch(Metadata.LeaderAndEpoch leaderAndEpoch) {
+        TopicPartition tp = TEST_PARTITION_1;
+        Node leader = LEADER_1;
+        OffsetResetStrategy strategy = OffsetResetStrategy.EARLIEST;
+        long offset = 5L;
+        Map<TopicPartition, OffsetAndTimestamp> expectedOffsets = Collections.singletonMap(tp,
+                new OffsetAndTimestamp(offset, 1L, leaderAndEpoch.epoch));
+        when(subscriptionState.partitionsNeedingReset(time.milliseconds())).thenReturn(Collections.singleton(tp));
+        when(subscriptionState.resetStrategy(any())).thenReturn(strategy);
+        expectSuccessfulRequest(Collections.singletonMap(tp, leader));
+
+        requestManager.resetPositionsIfNeeded();
+        assertEquals(1, requestManager.requestsToSend());
+
+        // Reset positions response with offsets
+        when(metadata.currentLeader(tp)).thenReturn(testLeaderEpoch(leader, leaderAndEpoch.epoch));
+        NetworkClientDelegate.PollResult pollResult = requestManager.poll(time.milliseconds());
+        NetworkClientDelegate.UnsentRequest unsentRequest = pollResult.unsentRequests.get(0);
+        ClientResponse clientResponse = buildClientResponse(unsentRequest, expectedOffsets);
+        clientResponse.onComplete();
+        assertTrue(unsentRequest.future().isDone());
+        assertFalse(unsentRequest.future().isCompletedExceptionally());
+    }
+
+    @Test
+    public void testResetPositionsThrowsPreviousException() {
```

Comment:
Does this test subsumes `testValidatePositionsFailureWithUnrecoverableAuthException`?

Reply:
They are definitely testing the same failure, but that failure is handled differently for the reset and validate. So with the test for the reset we end up testing the exception handling based on the `cachedResetPositionsException`, whereas with the test for validate we are testing the `cachedValidatePositionsException` handling. Each is set and retrieved in different functions for reset and validate, so making sure that both paths work as expected.

Reply:
Sorry, I left the comment on the wrong line. I was asking that `testValidatePositionsFailureWithUnrecoverableAuthException` and `testResetPositionsThrowsPreviousException` look very similar. Are they redundant?

Reply:
uhm those are the same two mentioned initially...do you mean a different one maybe?

Reply:
Ah, you are right. I thought `testResetPositionsThrowsPreviousException` is a validatePosition test because it follows a bunch of other validatePosition tests. Could we move this test to where other resetPosition tests are?

Reply:
Sure, moved it next to the reset related ones.